File: src/codes/exposed-complexity/ts/rfc_proxy.ts
interface UserProfile {
  load(userId: string): object;
}

// 本体
class UserProfileService implements UserProfile {
  load(userId: string): object {
    console.log(`プロフィール取得: ${userId}`);
    return { name: "Hiroshi", id: userId };
  }
}

// Proxy（認証と監査を担当）
class UserProfileProxy implements UserProfile {
  private real = new UserProfileService();
  private auth = new AuthService();
  private audit = new AuditService();

  load(userId: string): object {
    if (!this.auth.authenticate(userId)) {
      throw new Error("認証失敗");
    }
    const profile = this.real.load(userId);
    this.audit.log("プロフィール表示");
    return profile;
  }
}

// 利用例
const proxy = new UserProfileProxy();
const profile = proxy.load("user-123");
console.log(profile);

File: src/codes/exposed-complexity/py/rfc_proxy.py
from abc import ABC, abstractmethod

class UserProfile(ABC):
    @abstractmethod
    def load(self, user_id: str) -> dict:
        pass

class UserProfileService(UserProfile):
    def load(self, user_id: str) -> dict:
        print(f"プロフィール取得: {user_id}")
        return {"name": "Hiroshi", "id": user_id}

class UserProfileProxy(UserProfile):
    def __init__(self):
        self._real = UserProfileService()
        self._auth = AuthService()
        self._audit = AuditService()

    def load(self, user_id: str) -> dict:
        if not self._auth.authenticate(user_id):
            raise Exception("認証失敗")
        profile = self._real.load(user_id)
        self._audit.log("プロフィール表示")
        return profile

# 利用例
proxy = UserProfileProxy()
profile = proxy.load("user-123")
print(profile)
